lab/src/Microsoft.ServiceModel.Syndication/src/XmlReaderWrapper.cs

```diff
@@ -0,0 +1,575 @@
+﻿using System;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace Microsoft.ServiceModel.Syndication
+{
+    internal class XmlReaderWrapper : XmlReader
+    {
+        private XmlReader reader;
+
+        private Func<XmlReaderWrapper, Task<string>> getValueFunc;
+        private Func<XmlReaderWrapper, Task<string>> readElementStringFunc;
+        private Func<XmlReaderWrapper, Task<string>> readStringFunc;
+        private Func<XmlReaderWrapper, Task> readEndElementFunc;
+        private Func<XmlReaderWrapper, Task<XmlNodeType>> moveToContentFunc;
+        private Func<XmlReaderWrapper, string, string, Task> readStartElementFunc2;
+        private Func<XmlReaderWrapper, Task<bool>> isStartElementFunc;
+        private Func<XmlReaderWrapper, string, string, Task<bool>> isStartElementFunc2;
+        private Func<XmlReaderWrapper, Task> skipFunc;
+        private Func<XmlReaderWrapper, Task<bool>> readFunc;
+        private Func<XmlReaderWrapper, Task<string>> readInnerXmlFunc;
+
+        private XmlReaderWrapper(XmlReader reader)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            this.reader = reader;
+
+            if (this.reader.Settings.Async)
+            {
+                InitAsync();
+            }
+            else
+            {
+                Init();
+            }
+        }
+
+        public static XmlReaderWrapper CreateFromReader(XmlReader reader)
+        {
```

Comment:
This method should return an XmlReader. If the passed in XmlReader supported async, then we shouldn't wrap it. If it doesn't, then we should wrap it.